### PR TITLE
create a PR to develop PayPal Plus Payment method

### PR DESCRIPTION
**Pull request by caveman86**
correct and reopen the PR , and I tried to create a new PR in case old one should be deleted

### CODE REVIEW

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ */
+
+
+namespace Aimeos\MShop\Service\Provider\Payment;
+
+
+use Omnipay\Omnipay as OPay;
+use Aimeos\MShop\Order\Item\Base as Status;
+
+use Illuminate\Support\Facades\Auth;
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Please remove that because there must be no reference to Laravel

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Payment provider for payment gateways supported by the PaypalPlus library.
+ *
+ * @package MShop
+ * @subpackage Service
+ */
+class PaypalPlus
+extends    \Aimeos\MShop\Service\Provider\Payment\OmniPay
+//	extends \Aimeos\MShop\Service\Provider\Payment\Base
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Please remove the comment

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+		$base = $this->getOrderBase( $order->getBaseId(), $parts );
+		$data = $this->getData( $base, $order->getId(), $params );
+		$urls = $this->getPaymentUrls();
+
+		$data['cancelUrl'] = $this->getConfigValue( 'cancelUrl', '' ) ;
+
+		if(empty($data['cancelUrl'])) {
+			$data['cancelUrl'] = $_SERVER['HTTP_ORIGIN'] ;
+		}
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: The cancelUrl is incorrect. If it's required by paypal, it's the same as 'payment.url-success'. There, the customer can choose to retry or change the payment. The cancelUrl is always set so you can remove these lines:
https://github.com/aimeoscom/ai-payments/blob/79f292340c7642b8914c326722d4466c7c8c73d1/lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php#L719-L726

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		try
+		{
+			$response = $this->sendRequest( $order, $data );
+			$isSuccessful = $response->isSuccessful()  ;
+
+			$approval_url = "" ;
+			$testmode = "" ;
+			$countryid = "" ;
+
+			if( $isSuccessful )
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: You can shorten this to:
```
if( $response->isSuccessful() )
```

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				$this->saveRepayData( $response, $base->getCustomerId() );
+
+				$status = $this->getValue( 'authorize', false ) ? Status::PAY_AUTHORIZED : Status::PAY_RECEIVED;
+				$this->saveOrder( $order->setPaymentStatus( $status ) );
+				
+				if(!empty($response->getData()['links']['1']['href'])){
+					$approval_url = $response->getData()['links']['1']['href'] ;
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: We don't use variable names with underscores. Please use `$approvalUrl` instead.

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+				}
+		
+				
+				$testmode = $this->getConfigValue( 'testmode', '' ) ;
+				if(!empty($testmode) && $testmode=="1"){
+					$testmode = "sandbox" ;
+				}elseif(!empty($testmode) && $testmode=="0"){
+					$testmode = "live" ;
+				}
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: You can shorten this to:
```
$this->getConfigValue( 'testmode' ) ? 'sandbox' : 'live';
```
and move it to the `getPayPalPlusJs()` method

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	
+				$countryid = Auth::user()->countryid;
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Don't use Laravel code because the driver has to work without Laravel. You can get the country code using:
```
		$addresses = $base->getAddress( \Aimeos\MShop\Order\Item\Base\Address\Base::TYPE_PAYMENT );

		if( ( $address = current( $addresses ) ) !== false ) {
			$countryid = $address->getCountryId();
		}
```

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+
+		
+		return new \Aimeos\MShop\Common\Helper\Form\Standard( $urls['returnUrl'] ?? '', 'POST', [], false ,$this->getPayPalPlusJs($approval_url,$testmode ,$countryid ) );
+
+	}
+
+	protected function getPayPalPlusJs($approval_url,$testmode ,$countryid ) 
+	{
+		echo '
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: The `echo` is wrong, you need to `return` the string

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+		<script type="application/javascript">		  
+			var ppp = PAYPAL.apps.PPP({
+			"approvalUrl": approval_ur,
+			"placeholder": "ppplus",
+			"country":"' . $countryid . '" ,
+			"mode": "' . $testmode . '",
+			});
+		</script>
+		';
+		exit();
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: You are not allowed to terminate the whole application in a payment driver. Please remove the `exit()`.

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		var approval_ur = \'' . $approval_url . '\'; 
+		</script> 
+		<script type="application/javascript">		  
+			var ppp = PAYPAL.apps.PPP({
+			"approvalUrl": approval_ur,
+			"placeholder": "ppplus",
+			"country":"' . $countryid . '" ,
+			"mode": "' . $testmode . '",
+			});
+		</script>
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: You can combine the script tags and use the URL directly without creating a new variable.

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 3 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: but it keeps cancelUrl as returnUrl unless I give it value as in my code !

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 8 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: here is the problem that I have described previously in PR 27 .. 
the issue at js ppp library from Paypal as it demands to be loaded as landing page -unlike of stripe js library- ,so I should terminate the code to force it to stay till ppp is loaded completely .. I just replace die() with exit() .
 now I am spending my time to learn how to create new controller and its view -but unfortunately I don't get any resources about that- as the PaypalPlus is an part of new created extension

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 9 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: I described the problem in previous comment

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 3 above)
Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Usually, that's desired. The success and cancel URLs both point to checkout confirm page where the customer has the option to retry or change the payment (also to non-PayPal options)

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 8 above)
Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Doesn't the PayPal code embed into the existing page? Because it's JS code, it should be the case.
Do you have a link to the documentation?

Don't start creating a new controller because it's the wrong way.

**Comment by aimeoscom** [comment]
There are also coding style issues you have to resolve:
https://travis-ci.org/github/aimeoscom/ai-payments/jobs/749927993

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 8 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: > Doesn't the PayPal code embed into the existing page? Because it's JS code, it should be the case.
> Do you have a link to the documentation?
yes sure .. here it is : https://developer.paypal.com/docs/integration/paypal-plus/mexico-brazil/integrate-a-payment-selection-page/#display-mini-browser
further more , I have created a pure php and follow instructions and it worked fine ! so the problem raised when we use Aimeos ..
> Don't start creating a new controller because it's the wrong way.
I have started already walk thought as I want to use another option in Paypal Plus Js library which is "use popup window" , but I crashed with a lot of Aimeos coding way to be implemented , for that reason I have create a topic in Aimeos Help "https://aimeos.org/help/help-f15/new-independent-completely-extension-t3346.html" which it is a must for me if I will continue working on Aimeos , no matter if that required for this PR or not .. thanks in advance

Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: The docs say that you can add a button to open another browser window or use and iframe. Did you try those options?
Your method can contains arbitrary HTML and/or JS code so both should be possible.

Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: yes .. I tried .. 
but same problem .. which is the code doesn't wait ppp js library to be loaded .. it goes to next statements unless I exit() or die() the execution !
please help !

Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Use "onsite: yes" in the admin backend configuration for the payment option to stay at the page. Then, no automatically executed post by the Javascript should occur.

Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: I tried this option too and it fails as there are no way to inject required ppp js inside onsite view file  , it just show a credit card and expiration date and other input boxes which I don't need them ... 
but I will try it again

Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Your JS is always injected:
https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L138

You can overwrite this method in your class and return an empty array to avoid rendering any fields:
https://github.com/aimeoscom/ai-payments/blob/master/lib/custom/src/MShop/Service/Provider/Payment/Stripe.php#L117-L126

**Comment by caveman86** [comment]
Dear Norbert,
I hope you're doing well ..
firstly, thanks for your helping here and on GitHub and in aimeos.org/help ..
secondly .. I would like to suggest to allow me upload my solution -which is not in best way that you guide me on github to do- to your client ..
then, after I will be reached to best way to do the solution, I will update the code for your client again ...
thus , your client won't be annoyed for late as the code works fine at the end whatever its way to be coded
without make ypur client pay anything till I update the code second time ..
is that possible ?
Regards

________________________________

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 8 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: your idea was great , but if there are no fileds , it won't stay at that page and goes through confirmation page without waiting js to be loaded .. so, I have added one filed -firstname for example- and that worked !
but I need to handle this field and need to remove "pay now" button ... 
aa

Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Use a hidden field (public => false) to avoid the field.

We can also extend the template and add here: https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L167
```
<?php elseif( $this->get( 'standardMethod', 'POST' ) === '' ) : ?>
	<a class="btn btn-default btn-lg" href="<?= $enc->attr( $this->standardUrlPayment ); ?>">
		<?= $enc->html( $this->translate( 'client', 'Change payment' ), $enc::TRUST ); ?>
	</a>
<?php else : ?>
	<button class="btn btn-primary btn-lg btn-action">
		<?= $enc->html( $this->translate( 'client', 'Proceed' ), $enc::TRUST ); ?>
	</button>
<?php endif; ?>

Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: I tried it but unfortunately public => false is same as deleting it .. it makes execution keep run without wait .. 
anyway I will try working this point
thanks a lot

Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: If we change the template too, the JS won't POST the form any more. Just use
```
return new \Aimeos\MShop\Common\Helper\Form\Standard( '', '', true ,$this->getPayPalPlusJs($approval_url,$testmode ,$countryid ) );
```
in your payment provider.

Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: head shut , it worked :D

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 1 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: done

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 2 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: done

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 3 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: done , but how come that approved !
client will be forwarded to confirmation page and empty his basket ! 
any way .. I have commented out the code that set cancelUrl just to get sure from you about solving it

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 4 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: done

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 6 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: done

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 7 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: done .. 
but why should the Driver has to work without Laravel !?

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 10 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: I kept it to be more clearly as PayPal Plus documented it :)

**Comment by caveman86** [comment]
> There are also coding style issues you have to resolve:
> https://travis-ci.org/github/aimeoscom/ai-payments/jobs/749927993

I think Mr.Travis is satisfied now :)
Happy new year :)

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+use Aimeos\MShop\Order\Item\Base as Status;
+
+/**
+ * Payment provider for payment gateways supported by the PaypalPlus library.
+ *
+ * @package MShop
+ * @subpackage Service
+ */
+class PaypalPlus
+extends    \Aimeos\MShop\Service\Provider\Payment\OmniPay
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: There's a coding style issue, tab first then `extends`, then space, then the rest

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+		'address' => array(
+			'code' => 'address',
+			'internalcode'=> 'address',
+			'label'=> 'Send address to payment gateway too',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '0',
+			'required'=> false,
+		),
+		'authorize' => array(
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Does PayPalPlus support authorization? If no, this must be removed

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+			'code' => 'authorize',
+			'internalcode'=> 'authorize',
+			'label'=> 'Authorize payments and capture later',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '1',
+			'required'=> true,
+		),
+		'createtoken' => array(
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Remove because PayPalPlus doesn't support token based payment, at least not with the current implementation

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+			'label'=> 'Test mode without payments',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '0',
+			'required'=> true,
+		),
+	);
+	private $provider;
+
+	private $feConfig = array(
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Don't need this at all but we need the extension of the checkout process template

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+			$this->provider = OPay::create( 'PayPal_Rest' );
+			$this->provider->setTestMode( (bool) $this->getValue( 'testmode', false ) );
+			$this->provider->initialize( $this->getServiceItem()->getConfig() );
+		}
+
+		return $this->provider;
+	}
+
+
+	protected function sendRequest( \Aimeos\MShop\Order\Item\Iface $order, array $data ) : \Omnipay\Common\Message\ResponseInterface
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: This method seems to be unchanged, so you can remove it

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+
+		if( $this->getValue( 'createtoken', false ) ) {
+			$data['createCard'] = true;
+		}
+
+		if( $this->getValue( 'onsite', false ) || $this->getValue( 'address', false ) ) {
+			$data['card'] = $this->getCardDetails( $base, $params );
+		}
+
+		return $data + $this->getPaymentUrls();
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: You can shorten the whole method to:
```
return ['payerid' => $this->getContext()->getUserId() ?: $orderid] + parent::getData();
```

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+			$testmode = "" ;
+			$countryid = "" ;
+
+			if( $response->isSuccessful() )
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				$this->saveRepayData( $response, $base->getCustomerId() );
+
+				$status = $this->getValue( 'authorize', false ) ? Status::PAY_AUTHORIZED : Status::PAY_RECEIVED;
+				$this->saveOrder( $order->setPaymentStatus( $status ) );
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: This seems to be wrong here because the PayPalPlus form hasn't been rendered yet but you already set the order status to "authorized" or "received".

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+		else
+		{
+			$response = $provider->purchase( $data )->send();
+			$order->setPaymentStatus( Status::PAY_RECEIVED );
+		}
+
+		return $response;
+	}
+
+	protected function getPaymentForm( \Aimeos\MShop\Order\Item\Iface $order, array $params ) : \Aimeos\MShop\Common\Helper\Form\Iface
```
Comment (aimeoscom) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: The whole method seems to be merged from the original getPaymentForm() and processOrder() methods but I'm sure this is wrong because the processOrder() code should be called when the payment has been finished and the status is confirmed by the gateway.

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 11 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: for coding style .. 
I have set my vs code to to run phpcs and phpcbf , but as you see it still should improved .. 
so do you recommend any "standard" for phpcs to handle the style ?
or just use online tool like : https://www.tutorialspoint.com/online_php_formatter.htm .. ?
I will correct the rest of notes after testing them .. 
does

Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: This is the Aimeos PHPCS ruleset:
https://github.com/aimeos/aimeos-core/blob/master/misc/coding/Aimeos/ruleset.xml

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 17 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: yes it seems wrong , but it is working fine .. and I will explain that :
you are right that the status set as "received" before rendering of PPP , but that will be corrected by updateSync() which change the status to "unfinished" if the buyer cancel the PPP process , or keep it "received" if he complete the purchase successfully .. 
of course that updateSync() had worked just -as I think and not sure yet-  because of PPP returned urls like "cancelUrl and acceptedUrl ..etc" and not by normal execution workflow that other payment methods do .. (this point needs more work to be clear and optimized by me )
I have finished all other notes , but I didn't push the changes because I am still working on this point ..
so should I keep working on it , or return back to the last submitted code ?

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 18 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: answer of this issue is related to previous issue , I will return to it after solving previous one 
Regards

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 17 above)
Comment (aimeoscom) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: It's wrong nevertheless. If you choose PPP option, go to the process page an close the browser, the order is paid and you get the products for free.

Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: you are right , and sorry for delay .. 

I was working in this point and thanks God it is solved right now and everything are working as expected .. 

sure firstly I set the state of order is "unfinished" ..

the problem was that I couldn't get the result of outSite PPP JS process ...
solution : I have used $response of  $provider->completeAuthorize($params) .. 
BTY : I dissevered that our payerID is not accepted from PayPal !  ... and the conflict was because I missed that PayPal uses two phases -not one phase like other payment gateways!- process to complete the PPP payment ..
first to get approvalUrl , payerID and token ..  and this payerID should I use in second phase ! so I corrected them and no need for override getData() any more !
I have set payerID like so : $params['PayerID'] = $request->getQueryParams()['PayerID'] ; in updateSybc() 

then, everything are running as expected except two points : 
1. old checking ($response->getTransactionId() != $order->getId() )  not working ! as $response->getTransactionId() return null always ! and the solution is to get TransactionId from the request parameters as following : $response->getRequest()->getTransactionId()
2. method_exists($response, 'isCancelled') always null as it doesn't consider what happened after go out of Aimeos , so it consider the status of cancel during PPP process as "Refused" ! .. and I think it is wrong so I have add this concatenation like so : || ( !empty($response->getData()['name']) && $response->getData()['name'] == "PAYMENT_NOT_APPROVED_FOR_EXECUTION") and now it cancel during PPP is considered like "cancel" state ..

last thing , I have set PSR12 rules and corrected coding style I will correct anything in future after get your opinion of what I have mentioned previously 

thanks in advance ..

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+                $addresses = $base->getAddress(\Aimeos\MShop\Order\Item\Base\Address\Base::TYPE_PAYMENT);
+
+                if (($address = current($addresses)) !== false) {
+                    $countryid = $address->getCountryId();
+                }
+
+                if (empty($approvalUrl) || empty($countryid)) {
+                    throw new \Aimeos\MShop\Service\Exception($response->getMessage());
+                }
+            } elseif ($response->isRedirect()) {
```
Comment (aimeos) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php:121]: A redirect isn't possible, isn't it?

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+            }
+
+
+            if (method_exists($response, 'isSuccessful') && $response->isSuccessful()) {
+                $order->setPaymentStatus($status);
+            } elseif (method_exists($response, 'isPending') && $response->isPending()) {
+                $order->setPaymentStatus(Status::PAY_PENDING);
+            } elseif (
+                ( method_exists($response, 'isCancelled') && $response->isCancelled() )
+                || ( !empty($response->getData()['name']) && $response->getData()['name'] == "PAYMENT_NOT_APPROVED_FOR_EXECUTION")
```
Comment (aimeos) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php:206]: PAYMENT_NOT_APPROVED_FOR_EXECUTION sound like it's refused by PayPal, not cancelled by the user

Comment (aimeos) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php:206]: If you remove this line, is the rest of the method the same as the original one and can be removed?

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 17 above)
Comment (aimeos) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: The missing transaction ID is a weakness in Paypal_Rest driver and you should create a PR there to set the transaction ID from the request to the response. In the meantime, we can overwrite our updateSync() method and use `$response->getRequest()->getTransactionId() === $order->getId()`

Comment (aimeos) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: Same for `isCancelled()`. You should create a PR in the Omnipay Paypal_Rest driver that fixes that in the driver to return true for isCancelled() if PAYMENT_NOT_APPROVED_FOR_EXECUTION is returned by PayPal

**Comment by aimeos** [comment]
Please remember to create a PR in the aimeos/ai-client-html package which adds a branch if method is an empty string in the template, so there's no automatic POST to the confirmation page

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 17 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php]: ok , I will do that

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php (hunk 19 above)
Comment (caveman86) [reply lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php:121]: yes it is , but I kept it as I had copy updateSync() from onmipay.php ..
so ,no need for it and I will remove it

**Comment by caveman86** [comment]
> Please remember to create a PR in the aimeos/ai-client-html package which adds a branch if method is an empty string in the template, so there's no automatic POST to the confirmation page

sorry , but I didn't understand this note ..
I didn't add any code in ai-client-html !
which method do you mean ?

**Comment by aimeoscom** [comment]
We need an additional `elseif` here: https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L167
with:
```
<?php elseif( $this->get( 'standardMethod', 'POST' ) === '' ) : ?>
	<a class="btn btn-default btn-lg" href="<?= $enc->attr( $this->link( 'client/html/checkout/standard/url', ['c_step' => 'payment'] ) ); ?>">
		<?= $enc->html( $this->translate( 'client', 'Change payment' ), $enc::TRUST ); ?>
	</a>
```
to avoid the automatic redirect to the confirmation page.

**Comment by caveman86** [comment]
> We need an additional `elseif` here: https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L167
> with:
> 
> ```
> <?php elseif( $this->get( 'standardMethod', 'POST' ) === '' ) : ?>
> 	<a class="btn btn-default btn-lg" href="<?= $enc->attr( $this->link( 'client/html/checkout/standard/url', ['c_step' => 'payment'] ) ); ?>">
> 		<?= $enc->html( $this->translate( 'client', 'Change payment' ), $enc::TRUST ); ?>
> 	</a>
> ```
> 
> to avoid the automatic redirect to the confirmation page.

as we have used \Aimeos\MShop\Common\Helper\Form\Standard('', 'POST', [], true, $this->getPayPalPlusJs($approvalUrl, $countryid)); and the order state will not be as ""Received" till it is directed to successfully url after successfully process in PPP JS ... so I think there are no way to another checking !
anyway ..
I have reviewed the code from vendor/aimeos/aimeos-core/lib/mshoplib/src/MShop/Common/Helper/Form/Standard.php 
so nothing will make shop/checkout/process page to redirect to confirmation page ..

and even reviewed : vendor/aimeos/aimeos-core/lib/mwlib/src/MW/View/Standard.php that uses its functions to set and get parameters that has been set by each component's standard file such as ext/ai-client-html/client/html/src/Client/Html/Checkout/Standard/Process/Standard.php in our case .. 

so if we want to add another "if" ,we should customize and PR Standard.php at :
ext/ai-client-html/client/html/src/Client/Html/Checkout/Standard/Process/Standard.php to add a  new bool variable to certain value whenever payment method is "PayPal Plus" 
then , get this variable process-body-standard.php to prevent any redirection

but as I said , as we used return new \Aimeos\MShop\Common\Helper\Form\Standard('', 'POST', [], true, $this->getPayPalPlusJs($approvalUrl, $countryid)); as return in process page , there are no way to redirect to confirmation page ..
even no there are no calling to  getPayPalPlusJs function at all ..
am I right ?

**Comment by aimeoscom** [comment]
If we don't add another elseif, the Javascript will automatically submit the button (https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L169-L171) and forward the customer to the confirmation page. This was your initial problem you tried to solve with a die() if there's no FE-config added to PayPalPlus provider

**Comment by caveman86** [comment]
> If we don't add another elseif, the Javascript will automatically submit the button (https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L169-L171) and forward the customer to the confirmation page. This was your initial problem you tried to solve with a die() if there's no FE-config added to PayPalPlus provider

yeas I understand you now , but that resolved already by your suggestion in this link : 
https://github.com/aimeoscom/ai-payments/pull/28#discussion_r545842246

as I mentioned in last reply .. you recommended to add this line :
Aimeos\MShop\Common\Helper\Form\Standard('', 'POST', [], true);
instead of :
Aimeos\MShop\Common\Helper\Form\Standard( $url, 'POST', $list, false ); 

so after we removed $url and $list , form will not go anywhere

File: lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+
+            $params = (array) $request->getAttributes() + (array) $request->getParsedBody() + (array) $request->getQueryParams();
+            $params = $this->getData($base, $order->getId(), $params);
+            $params['transactionReference'] = $this->getTransactionReference($base);
+
+            if ($this->getValue('authorize', false) && $provider->supportsCompleteAuthorize()) {
+                $response = $provider->completeAuthorize($params)->send();
+                $status = Status::PAY_AUTHORIZED;
+            } elseif ($provider->supportsCompletePurchase()) {
+                $params['PayerID'] = $request->getQueryParams()['PayerID'] ;
```
Comment (aimeos) [inline lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php:185]: I would push this to getData() nevertheless because as soon as your PR for `$response->getTransactionId()` to return the order ID and PAYMENT_NOT_APPROVED_FOR_EXECUTION leads to "cancelled", we can remove updateSync() completely.

**Comment by aimeos** [comment]
For our coding style, we use more spaces than in PSR-2, e.g.
```
if( ( $value = $this->method( $arg1, arg2 ) ) !== null ) {
```
Can you please update the coding style accordingly?

**Comment by aimeos** [comment]
> so after we removed $url and $list , form will not go anywhere

OK, but we need to extend the template anyway because now a "Proceed" button is displayed with no functionality and this is annoying to users.
